packages/doenetml-worker-javascript/src/utils/dast/sugar.ts

```diff
@@ -0,0 +1,334 @@
+import { deepClone } from "@doenet/utils";
+import { ComponentInfoObjects } from "../componentInfoObjects";
+import {
+    ErrorRecord,
+    isSerializedAttribute,
+    isSerializedComponent,
+    PrimitiveAttributeValue,
+    SerializedComponent,
+    WarningRecord,
+} from "./types";
+import { findPreSugarIndsAndMarkFromSugar } from "../expandDoenetML";
+import {
+    isUnflattenedAttribute,
+    isUnflattenedComponent,
+} from "./intermediateTypes";
+import {
+    expandAllUnflattenedAttributes,
+    expandUnflattenedToSerializedComponents,
+} from "./convertNormalizdDast";
+import { convertToSerializedErrorComponent } from "./errors";
+
+/**
+ * If a component class has specified sugar instructions,
+ * then apply those instructors to the children,
+ * potentially creating new children and attributes.
+ */
+export function applySugar({
+    serializedComponents,
+    parentParametersFromSugar = {},
+    parentAttributes = {},
+    componentInfoObjects,
+    isAttributeComponent = false,
+}: {
+    serializedComponents: (string | SerializedComponent)[];
+    parentParametersFromSugar?: Record<string, any>;
+    parentAttributes?: Record<string, any>;
+    componentInfoObjects: ComponentInfoObjects;
+    isAttributeComponent?: boolean;
+}): {
+    components: (string | SerializedComponent)[];
+    errors: ErrorRecord[];
+    warnings: WarningRecord[];
+} {
+    let errors: ErrorRecord[] = [];
+    let warnings: WarningRecord[] = [];
+
+    const newComponents: (string | SerializedComponent)[] = [];
+
+    for (let component of serializedComponents) {
+        if (typeof component === "string") {
+            newComponents.push(component);
+            continue;
+        }
+
+        try {
+            let componentType = component.componentType;
+            let componentClass =
+                componentInfoObjects.allComponentClasses[componentType];
+            if (!componentClass) {
+                throw Error(`Unrecognized component type ${componentType}.`);
+            }
+
+            let componentAttributes: Record<string, PrimitiveAttributeValue> =
+                {};
+            // add primitive attributes to componentAttributes
+            for (let attrName in component.attributes) {
+                let attribute = component.attributes[attrName];
+                if (attribute.type === "primitive") {
+                    componentAttributes[attrName] = attribute.primitive;
+                }
+            }
+
+            let newParentParametersFromSugar = {};
+
+            let newComponent = component;
+
+            if (!component.skipSugar) {
```

Comment:
It'd be nicer to short circuit if `skipSugar` and avoid the extra indentation.

Reply:
except that we still recurse to children and attributes, below, within the loop, even if `skipSugar` is `true`